File: Codes/1993-operations-on-tree.py
'''
Author: LetMeFly
Date: 2023-09-23 09:26:03
LastEditors: LetMeFly
LastEditTime: 2023-09-23 09:54:53
'''
from typing import List

class LockingTree:

    def __init__(self, parent: List[int]):
        self.parent = parent
        self.lockUser = [0] * len(parent)
        self.child = [[] for _ in range(len(parent))]
        for i in range(1, len(parent)):
            self.child[parent[i]].append(i)


    def lock(self, num: int, user: int) -> bool:
        if self.lockUser[num]:
            return False
        self.lockUser[num] = user
        return True


    def unlock(self, num: int, user: int) -> bool:
        if self.lockUser[num] == user:
            self.lockUser[num] = 0
            return True
        return False


    def upgrade(self, num: int, user: int) -> bool:
        if not self.lockUser[num] and not self.__hasLockedParent__(num) and self.__hasLockedChildAndUnlock__(num):
            self.lockUser[num] = user
            return True
        return False
    

    def __hasLockedParent__(self, num: int) -> bool:
        while self.parent[num] != -1:
            num = self.parent[num]
            if self.lockUser[num]:
                return True
        return False
    

    def __hasLockedChildAndUnlock__(self, num: int) -> bool:
        hasLockedChild = False
        if self.lockUser[num]:
            self.lockUser[num] = 0
            hasLockedChild = True
        for thisChild in self.child[num]:
            hasLockedChild |= self.__hasLockedChildAndUnlock__(thisChild)
        return hasLockedChild


# Your LockingTree object will be instantiated and called as such:
# obj = LockingTree(parent)
# param_1 = obj.lock(num,user)
# param_2 = obj.unlock(num,user)
# param_3 = obj.upgrade(num,user)

File: Codes/1993-operations-on-tree.cpp
/*
 * @Author: LetMeFly
 * @Date: 2023-09-23 09:26:01
 * @LastEditors: LetMeFly
 * @LastEditTime: 2023-09-23 09:43:17
 */
#ifdef _WIN32
#include "_[1,2]toVector.h"
#endif

class LockingTree {
private:
    vector<int> lockUser;  // 谁锁的这个节点（0表示未锁）
    vector<int> parent;
    vector<vector<int>> child;

    bool hasLockedParent(int num) {
        while (parent[num] != -1) {
            num = parent[num];
            if (lockUser[num]) {
                return true;
            }
        }
        return false;
    }

    bool hasLockedChildAndUnlock(int num) {
        bool hasLockedChild = false;
        if (lockUser[num]) {
            lockUser[num] = 0;
            hasLockedChild = true;
        }
        for (int thisChild : child[num]) {
            hasLockedChild |= hasLockedChildAndUnlock(thisChild);
        }
        return hasLockedChild;
    }
    
public:
    LockingTree(vector<int>& parent): parent(parent) {
        lockUser = vector<int>(parent.size());
        child = vector<vector<int>>(parent.size());
        for (int i = 1; i < parent.size(); i++) {
            child[parent[i]].push_back(i);
        }
    }
    
    bool lock(int num, int user) {
        if (lockUser[num]) {
            return false;
        }
        lockUser[num] = user;
        return true;
    }
    
    bool unlock(int num, int user) {
        if (lockUser[num] == user) {
            lockUser[num] = 0;
            return true;
        }
        return false;
    }
    
    bool upgrade(int num, int user) {
        if (!lockUser[num] && !hasLockedParent(num) && hasLockedChildAndUnlock(num)) {
            lockUser[num] = user;
            return true;
        }
        return false;
    }
};

/**
 * Your LockingTree object will be instantiated and called as such:
 * LockingTree* obj = new LockingTree(parent);
 * bool param_1 = obj->lock(num,user);
 * bool param_2 = obj->unlock(num,user);
 * bool param_3 = obj->upgrade(num,user);
 */